### PR TITLE
UI: Task Group Scaling Controls

### PR DESCRIPTION
**Pull request by DingoEatingFuzz**
A fan favorite feature from Hashi UI is finally landing in the Nomad UI 🎉 

We have always been cautious about allowing casual changes to job definitions outside of the source control loop given our core principle of infrastructure as code. However! The new [scaling APIs](https://www.nomadproject.io/docs/job-specification/scaling/) means we can do the best of both worlds.

With this feature, in its initial form, if you have a task group that specifies a scaling block, the UI will let you scale a task group up and down within the min/max range declared in the job.

This can be done one of two ways:

1. **Quick action**: From the job detail page, task group rows now have +/- buttons when appropriate. Clicking these buttons will hit the scale endpoint and change the underlying count for the task group going through the expected deployment process. These buttons are debounced, so if you quickly click the + button five times, only one deployment with count+5 will be created.
2. **Precise action**: From the task group detail page, a task group can be scaled using +/- buttons just like in the table row, but it can also be scaled using manual text entry. Debouncing is used here as well.

Kapture 2020-06-18 at 23 42 21

image

### CODE REVIEW

File: ui/app/adapters/job.js [hunk 1]
```diff
@@ -68,4 +68,20 @@ export default class JobAdapter extends WatchableNamespaceIDs {
       },
     });
   }
+
+  scale(job, group, count, reason) {
+    const url = addToPath(this.urlForFindRecord(job.get('id'), 'job'), '/scale');
```
Comment (backspace) [inline ui/app/adapters/job.js:73]: I hadn’t known about `addToPath` before, I feel like I could have used it at some point, I’ll try to remember it exists 😯

File: ui/tests/pages/components/stepper-input.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  isPresent,
+  text,
+  triggerable,
+  value,
+} from 'ember-cli-page-object';
+
+export default scope => ({
+  scope,
+
+  isPresent: isPresent(),
```
[the remark this answers is not in the document]
Comment (DingoEatingFuzz) [reply ui/tests/pages/components/stepper-input.js:16]: 🤯 Today I learned!

File: ui/app/templates/jobs/job/task-group.hbs [hunk 3]
```diff
@@ -7,7 +7,21 @@
...
+    <div>
+      <Exec::OpenButton @job={{model.job}} @taskGroup={{model}} />
+      {{#if model.scaling}}
+      <StepperInput
+        data-test-task-group-count-stepper
+        @min={{model.scaling.min}}
+        @max={{model.scaling.max}}
+        @value={{model.count}}
+        @class="is-primary is-small"
+        @disabled={{or model.job.runningDeployment (cannot "scale job")}}
```
[the remark this answers is not in the document]
Comment (DingoEatingFuzz) [reply ui/app/templates/jobs/job/task-group.hbs:19]: Yeeeaaah. The reason (if you're curious) is there's a running deployment.

This definitely needs a tooltip, but it wasn't easy to shim it in with a good enough message, so I cut that corner to make the beta. I'll log it as a known issue though.

File: ui/tests/integration/stepper-input-test.js [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+    const input = find('[data-test-stepper-input]');
+
+    input.value = newValue;
+    assert.equal(StepperInput.input.value, newValue);
+
+    await StepperInput.input.esc();
+    assert.equal(StepperInput.input.value, this.value);
+  });
+});
```
[the remark this answers is not in the document]
Comment (DingoEatingFuzz) [reply ui/tests/integration/stepper-input-test.js:167]: I really appreciate you hammering on this component! I of course want the UX of all the components to be as good as possible.

I tried to keep as much of the default number input experience as possible, but I agree with your critique. I'll add the following to our known issues list:

1. invalid input should revert the value, not set to min (number input default)
2. clicking the label of the stepper input should focus the input field
3. when the input field is focused, the value should be highlighted so you don't have to deleting the existing value before providing a new one

File: ui/mirage/factories/task-group.js [hunk 5]
```diff
@@ -38,6 +40,31 @@ export default Factory.extend({
...
+          Min: 1,
+          Max: 5,
+          Policy: faker.random.boolean() && {
+            EvaluationInterval: '10s',
+            Cooldown: '2m',
+            Check: {
+              avg_conn: {
+                Source: 'prometheus',
+                Query:
+                  'scalar(avg((haproxy_server_current_sessions{backend="http_back"}) and (haproxy_server_up{backend="http_back"} == 1)))',
```
[the remark this answers is not in the document]
Comment (DingoEatingFuzz) [reply ui/mirage/factories/task-group.js:55]: Right? I copied it straight from the autoscaler docs.

**Comment by jippi** [comment]
Nice, excited to see hashi-ui becoming less and less relevant!

**Pull request merged**